packages/core/src/build/schema.ts

```diff
@@ -0,0 +1,161 @@
+import { BuildError } from "@/common/errors.js";
+import { type Schema, isPgEnumSym } from "@/drizzle/index.js";
+import { getSql } from "@/drizzle/kit/index.js";
+import { getTableColumns, is } from "drizzle-orm";
+import {
+  type PgEnum,
+  PgSchema,
+  PgSequence,
+  PgTable,
+  PgView,
+  getTableConfig,
+} from "drizzle-orm/pg-core";
+
+export const buildSchema = ({
+  schema,
+  instanceId,
+}: { schema: Schema; instanceId: string }) => {
+  const statements = getSql(schema, instanceId);
+
+  // find and validate namespace
+
+  let namespace: string;
+
+  for (const maybeSchema of Object.values(schema)) {
+    if (is(maybeSchema, PgSchema)) {
+      namespace = maybeSchema.schemaName;
+      break;
+    }
+  }
+
+  if (namespace! === undefined) {
+    namespace = "public";
+  }
+
+  for (const [name, s] of Object.entries(schema)) {
+    if (is(s, PgTable)) {
+      if (namespace === "public" && getTableConfig(s).schema !== undefined) {
+        throw new Error(
+          `Schema validation failed: All entities must use the same schema and ${name} uses a different schema '${getTableConfig(s).schema}' than '${namespace}'.`,
+        );
+      }
+      if (namespace !== "public" && getTableConfig(s).schema !== namespace) {
+        throw new Error(
+          `Schema validation failed: All entities must use the same schema and ${name} uses a different schema '${getTableConfig(s).schema ?? "public"}' than '${namespace}'.`,
+        );
+      }
+
+      let hasPrimaryKey = false;
+
+      for (const [columnName, column] of Object.entries(getTableColumns(s))) {
+        if (column.primary) {
+          if (hasPrimaryKey) {
+            throw new Error(
+              `Schema validation failed: '${name}' has multiple primary keys.`,
+            );
+          } else {
+            hasPrimaryKey = true;
+          }
+        }
+
+        if (column.isUnique) {
+          throw new Error(
+            `Schema validation failed: '${name}.${columnName}' has a unique constraint and unique constraints are unsupported.`,
+          );
+        }
+
+        if (column.generated !== undefined) {
+          throw new Error(
+            `Schema validation failed: '${name}.${columnName}' is a generated column and generated columns are unsupported.`,
```

Comment:
Yes

- serial: allowed
- default: allowed
- generated: not allowed
- generated identity: deprecated from drizzle?
- autoincrementing: I think this the same as serial?

Reply:
Here are the Drizzle docs on identity columns (https://orm.drizzle.team/docs/column-types/pg#identity-columns). Pretty much every resource online recommends using identity columns over serial columns. AFAICT both serial and identity use a sequence under the hood, so IMO we should support either both or neither.

Considering the issue where serial column values are returned as undefined by the in-memory store, I'm leaning towards neither, at least for the initial release. Happy to be convinced otherwise though.

Reply:
Nice. I do think that the concept of auto-incrementing primary key is actually super useful for ponder apps, specifically to replace `log.id`. Could we make an exception for tables that use `serial` or `generatedAsIdentity` as a primary key?

We could try to mirror the value in the in-memory store, it just might add a ton of complexity for not much gain.

Reply:
Yea, I agree that it's very useful and ideally we can add it. But on balance I think it's better to ship without, and add a robust solution rather than add something to the public API that could cause bugs / we might ultimately have to remove (if we can't come up with a robust solution).

I guess TL;DR my opinion is that the current approach where column values backed by sequences are not available in userland is a dealbreaker - it seems "astonishing" and divergent from SQL behavior.

Reply:
makes sense, i'll remove `serial`